src/ParamPassing.cc

```diff
@@ -0,0 +1,171 @@
+/*
+ * Copyright 2020 Open Source Robotics Foundation
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ *
+ */
+#include <cstring>
+#include <vector>
+
+#include "ParamPassing.hh"
+
+namespace sdf
+{
+inline namespace SDF_VERSION_NAMESPACE {
+
+//////////////////////////////////////////////////
+void updateParams(const tinyxml2::XMLElement *_childXmlParams,
+                  SDFPtr _includeSDF, Errors &_errors)
+{
+  // loop through <experimental:params> children
+  const tinyxml2::XMLElement *childElemXml = nullptr;
+  for (childElemXml = _childXmlParams->FirstChildElement();
+       childElemXml;
+       childElemXml = childElemXml->NextSiblingElement())
+  {
+    // element identifier
+    const char *childElemId = childElemXml->Attribute("name");
+    if (!childElemId)
+    {
+      tinyxml2::XMLPrinter printer;
+      childElemXml->Accept(&printer);
+
+      _errors.push_back({ErrorCode::ATTRIBUTE_MISSING,
+        "Missing element identifier. Skipping element modification:\n"
+        + std::string(printer.CStr())
+      });
+      continue;
+    }
+
+    std::string actionStr;
+    if (childElemXml->Attribute("action"))
+      actionStr = std::string(childElemXml->Attribute("action"));
+
+    // get element pointer to specified element using element identifier
+    ElementPtr  elem;
+    if (actionStr == "add")
+    {
+      // TODO(jenn) implement this, skipping for now
+      continue;
+    }
+    else
+    {
+      elem = getElementById(_includeSDF, childElemId,
+                            childElemXml->Name(), true);
+    }
+
+    if (elem == nullptr)
+    {
+      tinyxml2::XMLPrinter printer;
+      childElemXml->Accept(&printer);
+
+      _errors.push_back({ErrorCode::ELEMENT_MISSING,
+        "Could not find element <" + std::string(childElemXml->Name())
+        + " name='" + childElemXml->Attribute("name") + "'>. " +
+        "Skipping element modification:\n" + printer.CStr()
+      });
+      continue;
+    }
+
+    // TODO(jenn) element modifications
+  }
+}
+
+//////////////////////////////////////////////////
+ElementPtr getElementById(const SDFPtr _sdf,
+                          const std::string &_elemId,
+                          const std::string &_elemName,
+                          const bool &_prefixModelName)
+{
+  std::string modelName
+    = _sdf->Root()->GetFirstElement()->GetAttribute("name")->GetAsString();
```

Comment:
This assumes that the `GetAttribute("name")` returns a valid pointer. Is that a safe assumption?

Reply:
Yes, this looks for the `//model` name attribute from the included file and a `//model` requires a name. If there is no name then [this error will be encountered](https://github.com/osrf/sdformat/blob/eba07eb93136c460d0c6e43e0201da4b7af0c23c/src/parser.cc#L909-L911)